include/stx/formatter.h

```diff
@@ -0,0 +1,251 @@
+#pragma once
+
+#include <string>
+#include <string_view>
+#include <optional>
+#include <algorithm>
+#include <charconv>
+#include <type_traits>
+
+#include "string.h"
+
+namespace stx
+{
+
+/**
+ * Non-virtual formatter base for sharing functionality.
+ */
+struct formatter_base
+{
+    std::optional<char> fillc;
+    std::optional<char> justify;
+    std::optional<size_t> min_width;
+
+    /**
+     * Eat the first char from string_view `sv` if contained in `in`
+     * or if `in` is unset.
+    */
+    static std::optional<char> eat_char(std::string_view& sv, std::string_view in)
+    {
+        if (sv.empty())
+            return {};
+
+        auto c = sv.front();
+        if (!in.empty() && in.find(c) == std::string_view::npos)
+            return {};
+
+        sv.remove_prefix(1);
+        return c;
+    }
+
+    /**
+     * Eat the first number of type _T from string_view `sv`
+     */
+    template <class _T>
+    static std::optional<_T> eat_number(std::string_view& sv)
+    {
+        if (sv.empty())
+            return {};
+
+        _T v;
+        /* Msvc needs this super stupic iter->ptr conversion. */
+        auto begin = &(*sv.begin());
+        auto r = std::from_chars(begin, begin + sv.size(), v);
+        if (r.ptr == begin)
+            return {};
+
+        sv.remove_prefix(r.ptr - begin);
+        return v;
+    }
+
+    /**
+     * Formatter base ctor
+     *
+     * Parsing all generic/universal formatting options.
+     * Child classes must call justify_pre and justify_post
+     * before and after writing their values.
+     */
+    formatter_base(std::string_view& fmt)
+    {
+        /* {:[.][<>^][0-9]...} */
+        if (fmt.find_first_of("<^>") == 1u) {
+            fillc = eat_char(fmt, {});
+            justify = eat_char(fmt, "<^>");
+        /* {:[<>^][0-9]} */
+        } else if (fmt.find_first_of("<^>") == 0u) {
+            justify = eat_char(fmt, "<^>");
+        }
+
+        /* {:[0-9]} */
+        min_width = eat_number<unsigned>(fmt);
+
+        /* {:[.]} */
+        if (!fillc && !min_width)
+            fillc = eat_char(fmt, {});
+    }
+
+    template <class _Iter>
+    void justify_pre(size_t width, _Iter out)
+    {
+        if (justify == '>') {
+            if (width < min_width && min_width)
+                std::fill_n(out, *min_width - width, fillc.value_or(' '));
+        } else if (justify == '^') {
+            if (width < min_width)
+                std::fill_n(out, (*min_width - width) / 2, fillc.value_or(' '));
+        } else if (justify == '<') {
+            /* Special case: Prepend 1 fillc if value is non-empty.
```

Comment:
Technically the trick should also work with right-justify, no? But it's a tiny limitation.

Reply:
It does, but on the right side, see the next function.